pkg/config/msp.go

```diff
@@ -0,0 +1,142 @@
+func AddRootCAToMSP(rootCA *x509.Certificate, orgName string, config *cb.Config) error {
+	if (rootCA.KeyUsage & x509.KeyUsageCertSign) == 0 {
+		return errors.New("certificate KeyUsage must be x509.KeyUsageCertSign")
+	}
+	if !rootCA.IsCA {
+		return errors.New("certificate must be a CA certificate")
+	}
+
+	org := getApplicationOrg(config, orgName)
+	fabricMSPConfig, err := getFabricMSPConfig(org)
+	if err != nil {
+		return fmt.Errorf("getting msp config: %v", err)
+	}
+
+	buffer := bytes.NewBuffer(nil)
+	err = pem.Encode(buffer, &pem.Block{Type: "CERTIFICATE", Bytes: rootCA.Raw})
+	if err != nil {
+		return err
+	}
+	fabricMSPConfig.RootCerts = append(fabricMSPConfig.RootCerts, buffer.Bytes())
+
+	err = addMSPConfigToOrg(org, fabricMSPConfig)
+	if err != nil {
+		return fmt.Errorf("adding msp config to org: %v", err)
+	}
+
+	return nil
+}
+
+func getFabricMSPConfig(org *cb.ConfigGroup) (*mb.FabricMSPConfig, error) {
+	configValue := org.Values[MSPKey]
+
+	mspConfig := &mb.MSPConfig{}
+	err := proto.Unmarshal(configValue.Value, mspConfig)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshalling mspConfig: %v", err)
+	}
+
+	fabricMSPConfig := &mb.FabricMSPConfig{}
+	err = proto.Unmarshal(mspConfig.Config, fabricMSPConfig)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshalling mspConfig: %v", err)
+	}
+
+	return fabricMSPConfig, nil
+}
+
+func addMSPConfigToOrg(org *cb.ConfigGroup, fabricMSPConfig *mb.FabricMSPConfig) error {
+	configValue := org.Values[MSPKey]
+
+	mspConfig := &mb.MSPConfig{}
+	err := proto.Unmarshal(configValue.Value, mspConfig)
+	if err != nil {
+		return fmt.Errorf("unmarshalling mspConfig: %v", err)
+	}
+
+	serializedMSPConfig, err := proto.Marshal(fabricMSPConfig)
+	if err != nil {
+		return fmt.Errorf("marshalling updated mspConfig: %v", err)
+	}
+
+	mspConfig.Config = serializedMSPConfig
+
+	err = addValue(org, mspValue(mspConfig), AdminsPolicyKey)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// RevokeCertificateFromMSP takes an x509 certificate and adds it to the
+// revocation list for the specified application org MSP.
+func RevokeCertificateFromMSP(cert *x509.Certificate, orgName string, config *cb.Config) error {
+	org := getApplicationOrg(config, orgName)
+	fabricMSPConfig, err := getFabricMSPConfig(org)
+	if err != nil {
+		return fmt.Errorf("getting msp config: %v", err)
+	}
+
+	// TODO validate that this certificate was issued by this MSP
+
+	if len(fabricMSPConfig.RevocationList) == 0 {
+		// TODO create a CRL here if one doesn't exist?
+		return errors.New("no CRL in RevocationList")
+	}
+
+	// just parse the first CRL and add it there for now
+	crl, err := x509.ParseCRL(fabricMSPConfig.RevocationList[0])
+	if err != nil {
+		return fmt.Errorf("parsing CRL: %v", err)
+	}
+
+	crl.TBSCertList.RevokedCertificates = append(crl.TBSCertList.RevokedCertificates,
+		pkix.RevokedCertificate{
+			SerialNumber: cert.SerialNumber,
+			// RevocationTime: time.Now(),
```

Comment:
Why is this commented out?

Reply:
It was on purpose because I thought the ASN.1 marshaled bytes would change and break the test if we used the current time. Turns out it doesn't for some reason so I'll uncomment it. :)